oeps/oep-0019-bp-django-caches.rst

```diff
@@ -0,0 +1,228 @@
+   collisions in a global space like Memcached.
+
+Note: When and if caching is warranted in any particular circumstance is beyond
+the scope of this OEP.
+
+Decision
+--------
+
+Common Caching Defect and Fix
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The following is a common defect when caching using a Django caches::
+
+    cached_value = cache.get(key)
+    if cached_value:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # be treated as a cache miss above.
+    return cached_value
+
+The simplest fix for this defect is::
+
+    _CACHE_MISS = object()
+
+    cached_value = cache.get(key, _CACHE_MISS)
+    if cached_value is _CACHE_MISS:
+        # calculate value, set in cache, and return
+        # if the calculated value is Falsey (e.g., None or []), it will
+        # no longer be treated as a cache miss above.
+    return cached_value
+
+Use libraries that makes it simpler to avoid the above defect. See
+`Implementations`_ for more details.
+
+Tiered Cache with Request Caching
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Use a library that allows tiered-caching calls that use a Request Cache backed
+by Memcached. Multiple calls for the same value within a request will not
+require additional network calls to Memcached.
+
+See `Implementations`_ for more details.
+
+Force Cache Miss
+~~~~~~~~~~~~~~~~
+
+Ensure that new caching functionality has a simple method by which to force a
+cache miss and thus force a cached value to be recalculated. This is useful
+both for testing purposes, as well as for business purposes if a problem occurs
+in Production that requires a capability for refreshing a cached value.
+
+Cache Decorators
+~~~~~~~~~~~~~~~~
+
+The following recommendation requires additional assessment.
+
+When caching, create cache methods where the boilerplate caching code is hidden
+behind a decorator. This practice allows the more important business logic to
+stand out.
+
+Note that the boilerplate logic hidden behind the decorator must include
+opinionated code regarding how keys are defined and when and if the key will
+change. More assessment is required here to determine the recommended
+decorator(s) to use.
+
+Sample Before::
+
+    def get_my_business_value(x):
+        cached_value = cache.get(key, _CACHE_MISS)
+        if cached_value is _CACHE_MISS:
+            new_value = calculate_my_business_value()
+            key = get_my_business_key(x)
+            cache.set(key, new_value)
+            return new_value
+        return cached_value
+
+Sample After::
+
+    @sample_cache_decorator(['x'])
+    def get_my_business_value(x):
+        # calculate my_business_value
+        ...
+        return my_business_value
+
+See `Implementations`_ for more details.
+
+Generating Cache Keys
+~~~~~~~~~~~~~~~~~~~~~
+
+Either use a cache decorator that generates the cache keys, or use a trusted
```

Comment:
Does it _have_ to be a decorator?  What's wrong with a private helper method like:
```python
def _cache_key(self, x):
    return 'my_module.my_class.my_function.{}'.format(x)
```

Reply:
Will make this more clear.